Several fixes to support long-lived persistent connections

tgraf opened the pull request.

manalibhutiyani reviewed (approved):
Approving for Kafka bits, they look fine.

rlenglet commented on pkg/proxy/dialer.go:
```diff
@@ -66,6 +81,11 @@ func ciliumDialer(identity int, network, address string) (net.Conn, error) {
 		return nil, fmt.Errorf("unable to connect: %s", err)
 	}
 
+	if err := setKeepAlive(c); err != nil {
```
We also need to set those options on request connections, in `pkg/proxy/socket.go`'s `proxySocket.Accept`.

tgraf replied:
fixed

rlenglet commented on pkg/proxy/dialer.go:
```diff
@@ -66,6 +81,11 @@ func ciliumDialer(identity int, network, address string) (net.Conn, error) {
 		return nil, fmt.Errorf("unable to connect: %s", err)
 	}
 
+	if err := setKeepAlive(c); err != nil {
```
Shouldn't we also call `TCPConn.SetLinger` with a delay consistent with `proxyConnectionCloseTimeout`, in order to bound the time it takes to close a connection?
Again, we'd need to do that on request connections as well.

rlenglet replied:
I'll set the SO_LINGER option on the request connection in https://github.com/cilium/cilium/pull/2860. I think that's required at least on request side.

rlenglet commented on bpf/lib/common.h:
```diff
@@ -398,7 +398,10 @@ struct ct_state {
+/* Lifetime of a proxy redirection entry is an entire day. All proxies should
+ * be using TCP keepalive to force some traffic over the connection
+ * periodically. */
+#define PROXY_DEFAULT_LIFETIME 86400
```
Does this lifetime need to be that long? With https://github.com/cilium/cilium/pull/2860, we'll have a guarantee that Kafka connections don't live longer than ~1 minute after a redirect is removed. I don't understand yet the behavior of Envoy, but we can probably set a shorter lifetime.

tgraf replied:
I'm happy to adjust this down again once we understand the behaviour of Envoy.

tgraf committed:
proxy: Cleanup BPF proxy map on removal of redirect

Entries will still expire but this accelerates the removal which is required in
combination with the extended lifetime.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
kafka: Report error message when writing access log to debug log

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
cilium: Add identity and lifetime to cilium bpf proxy list

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Always update the lifetime of proxymap entries

The proxymap insertion logic depends on the NEEDS_UPDATE define to decide
whether the lifetime fields should be filled. NEEDS_UPDATE is defined if
the kernel does not support LRU maps. Unfortunately, the proxymaps are not
converted to LRU maps even if available which caused the lifetime not to
be updated on kernels which support LRU maps.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Increase proxymap entry expire time to 1d

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Keep connection tracking entries around for at least 12h

Signed-off-by: Thomas Graf <thomas@cilium.io>
